crates/rover-client/src/blocking/client.rs

```diff
@@ -50,80 +58,113 @@ impl Client {
                     }
-
-                    return Err(RoverClientError::GraphQl {
-                        msg: errs
-                            .into_iter()
-                            .map(|err| err.message)
-                            .collect::<Vec<String>>()
-                            .join("\n"),
-                    });
+                    status_code => Err(RoverClientError::ClientError {
+                        msg: status_code.to_string(),
+                    }),
                 }
-
-                if let Some(data) = response_body.data {
-                    Ok(data)
+            }
+            Err(e) => {
+                if response_status.is_success() {
+                    Err(e.into())
                 } else {
-                    Err(RoverClientError::MalformedResponse {
-                        null_field: "data".to_string(),
+                    Err(RoverClientError::ClientError {
+                        msg: response_status.to_string(),
                     })
                 }
             }
-            // This block specifically handles an error that is returned when
-            // Introspection is set to false on a production ApolloServer.
-            //
-            // We first check for a 400 HTTP Status Code (Bad Request). We then
-            // get the message sent by the server and display that to our users.
-            StatusCode::BAD_REQUEST => {
-                // It's not a given that an HTTP response is valid JSON,
-                // so let's match for a successful parse. Return a standard 400
-                // RoverClientError if we are unable to parse.
-                match response.json::<graphql_client::Response<Q::ResponseData>>() {
-                    Ok(body) => {
-                        if let Some(errs) = body.errors {
-                            return Err(RoverClientError::ClientError {
-                                msg: errs[0].message.to_string(),
-                            });
-                        }
-                        Err(RoverClientError::ClientError {
-                            msg: StatusCode::BAD_REQUEST.to_string(),
-                        })
-                    }
-                    Err(_) => Err(RoverClientError::ClientError {
-                        msg: StatusCode::BAD_REQUEST.to_string(),
-                    }),
-                }
-            }
-            status => Err(RoverClientError::ClientError {
-                msg: status.to_string(),
-            }),
         }
     }
 }
+
+fn handle_graphql_body_errors(errors: Vec<GraphQLError>) -> Result<(), RoverClientError> {
+    if errors.is_empty() {
+        Ok(())
+    } else if errors[0].message.contains("406") {
+        Err(RoverClientError::MalformedKey)
+    } else {
+        Err(RoverClientError::GraphQl {
+            msg: errors
+                .into_iter()
+                .map(|error| error.message)
+                .collect::<Vec<String>>()
+                .join("\n"),
```

Comment:
Any errors in the response body that we don't special case are then returned as part of this. As for unauthenticated users, I think that might even be handled elsewhere, though I could be mistaken.

```console
$ APOLLO_KEY="bad" cargo run -- config whoami  
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/rover config whoami`                
Checking identity of your API key against the registry.
error[E014]: The API key you provided is malformed. An API key must have three parts separated by a colon.
              Try running `rover docs open api-keys` for more details on Apollo's API keys.
```

```console
$ APOLLO_KEY="user:gh.EverlastingBugstopper:1231231231231231231231" cargo run -- config whoami 
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/rover config whoami`
Checking identity of your API key against the registry.
error[E013]: The registry did not recognize the provided API key
              Check your API key to make sure it's valid (are you using the right profile?).
```

Reply:
Ah, yea this line should do it, great!